### PR TITLE
Add Webhooks in store's settings

### PR DESCRIPTION
**Pull request by NicolasDorier**
This cover the new webhooks as alternative to #2025
The greenfield API not done yet, this only cover the UI.

Features are:

* Using standard HMAC (github like) for secure callback
* Ability to chose which callback to subscribe to in a granular maner for a single webhook
* Ability to browse the delivery (as well as the request)
* Ability to redeliver manually
* In-order delivery at the webhook level, if a delivery fails and automatic redelivery is turned on, the other delivery won't happen until the previous succeed or failed all redeliveries.

The UX follow github's one and the UI is copying the look & feel of @dennisreimann latest changes to the store.

The design does not couple webhooks to stores specifically, meaning we can reuse that later for events at the server or user level.

image

image

### CODE REVIEW

**Comment by NicolasDorier** [comment]
Added the deliveries also at invoice level

image

Delivery is forbidden to Guests

**Review by Kukks — CHANGES_REQUESTED** [review]
* Needs integration in GreenField stores update endpoint
* Needs endpoints for invoices and stores in GreenField to view deliveries
* Needs docs on how to verify in GreenField
* Needs docs for all different webhook events
* Needs a way for invoices to be able to specify additional webhook subscriptions on creation

IMO this version over #2025  overcomplicates and adds waaaay more stuff.

File: BTCPayServer/HostedServices/WebhookNotificationManager.cs [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+                    return WebhookEventType.InvoiceCreated;
+                case InvoiceEventCode.Expired:
+                case InvoiceEventCode.ExpiredPaidPartial:
+                    return WebhookEventType.InvoiceExpired;
+                case InvoiceEventCode.FailedToConfirm:
+                case InvoiceEventCode.MarkedInvalid:
+                    return WebhookEventType.InvoiceInvalid;
+                case InvoiceEventCode.PaidInFull:
+                    return WebhookEventType.InvoicePaidInFull;
+                case InvoiceEventCode.ReceivedPayment:
```
[the remark this answers is not in the document]
Comment (NicolasDorier) [reply BTCPayServer/HostedServices/WebhookNotificationManager.cs:180]: Yeah I did not wanted to include payment for now for this reason.

File: BTCPayServer.Data/Data/ApplicationDbContext.cs [hunk 2]
```diff
@@ -63,6 +63,11 @@ public DbSet<RefundData> Refunds
         public DbSet<U2FDevice> U2FDevices { get; set; }
         public DbSet<NotificationData> Notifications { get; set; }
 
+        public DbSet<StoreWebhookData> StoreWebhooks { get; set; }
```
[the remark this answers is not in the document]
Comment (NicolasDorier) [reply BTCPayServer.Data/Data/ApplicationDbContext.cs:66]: We have a limited amount of scopes, so I think that's fine.
This is an important feature that I expect will be used a lot by integrators.

Also other scopes don't have to use table if they don't want to. For example, the webhooks ids of server or app could be saved into blobs instead.

Arguably, I could have done this for store... maybe I should.

[the remark this answers is not in the document]
Comment (NicolasDorier) [reply BTCPayServer.Data/Data/ApplicationDbContext.cs:66]: I will try to remove the invoice and store table. It may fit into blobs.

**Comment by NicolasDorier** [comment]
For greenfield, I want to do it in a separate PR.

File: BTCPayServer/HostedServices/WebhookNotificationManager.cs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+            _InvoiceEventsByWebhookId.Add(context.WebhookId, channel);
+            channel.Writer.TryWrite(context);
+            _ = Process(context.WebhookId, channel);
+        }
+
+        private WebhookEventType? GetWebhookEvent(InvoiceEventCode eventCode)
+        {
+            switch (eventCode)
+            {
+                case InvoiceEventCode.Completed:
```
[the remark this answers is not in the document]
Comment (NicolasDorier) [reply BTCPayServer/HostedServices/WebhookNotificationManager.cs:165]: For completed, it should just be hidden. There is no reason to leak this information anywhere.

For other stuff like reason why the invoice is invalid, I will just add event specific fields I think.

**Comment by NicolasDorier** [comment]
@Kukks added all your points except

> Needs a way for invoices to be able to specify additional webhook subscriptions on creation

Which I don't think is necessary. Bitpay API allowed that, but there is no reason to keep this, as everybody can instead just register their webhook at the store level.

**Pull request merged**